src/policy/packages.cpp

```diff
@@ -147,3 +147,78 @@ bool IsChildWithParentsTree(const Package& package)
         return true;
     });
 }
+
+// Calculates curr_tx's in-package ancestor set. If the tx spends another tx in the package, calls
+// visit() for that transaction first, since any transaction's ancestor set includes its parents'
+// ancestor sets. Transaction dependency cycles are not possible without breaking sha256 and
+// duplicate transactions were already checked, so this won't recurse infinitely.  After this
+// function returns, curr_tx's entry in ancestor_set_map is guaranteed to contain a non-empty
+// ancestor_subset.
+void visit(const CTransactionRef& curr_tx,
+           const std::map<Txid, CTransactionRef>& txid_to_tx,
+           std::map<Txid, std::set<Txid>>& ancestor_set_map)
+{
+    const Txid& curr_txid = curr_tx->GetHash();
+    // Already visited? Return now.
+    auto curr_result_iter = ancestor_set_map.find(curr_txid);
+    if (curr_result_iter == ancestor_set_map.end()) return;
```

Comment:
`if (curr_result_iter == ancestor_set_map.end()) return;` is redundant the map has already been populated with all package transactions, the statement will always be false.
So assert for that.
And return early when the transaction in-package ancestor set is not empty, that means its has been visited
```suggestion
    auto curr_result_iter = ancestor_set_map.find(curr_txid);
    // All transactions were populated to ancestor_set_map this should never happen
    assert(curr_result_iter != ancestor_set_map.end());
    // Already visited? Return now.
    if (!curr_result_iter->second.empty()) return;
```

Reply:
Yep good catch! I've now built this into `PackageV3Checks` instead of a separate function, so marking this as resolved.